src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,351 @@
+                    (*err != CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute)));
+        return *err;
+    }
+
+    // All the global attributes that we do not have metadata for are
+    // read-only (i.e. cannot write atribute_list/event_list/accepted_cmds/generated_cmds)
+    //
+    // so if no metadata available, we wil lreturn an error
+    const EmberAfAttributeMetadata ** attributeMetadata = std::get_if<const EmberAfAttributeMetadata *>(&metadata);
+    if (attributeMetadata == nullptr)
+    {
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+    }
+
+    if ((*attributeMetadata)->IsReadOnly() && !request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        // Internal is allowed to try to bypass read-only updates, however otherwise we deny read-only
+        // updates
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedWrite);
+    }
+
+    if ((*attributeMetadata)->MustUseTimedWrite() && !request.writeFlags.Has(InteractionModel::WriteFlags::kTimed))
+    {
+        return CHIP_IM_GLOBAL_STATUS(NeedsTimedInteraction);
+    }
+
+    if (request.path.mDataVersion.HasValue())
+    {
+        std::optional<InteractionModel::ClusterInfo> clusterInfo = GetClusterInfo(request.path);
+        if (!clusterInfo.has_value())
+        {
+            ChipLogError(DataManagement, "Unable to get cluster info for Endpoint %x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+
+        if (request.path.mDataVersion.Value() != clusterInfo->dataVersion)
+        {
+            ChipLogError(DataManagement, "Write Version mismatch for Endpoint %x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+    }
+
+    std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(
+        request.path, GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId), decoder);
+    ReturnErrorCodeIf(aai_result.has_value(), *aai_result);
+
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, gEmberAttributeIOBufferSpan));
+
+    EmberAfAttributeSearchRecord record;
+    record.endpoint    = request.path.mEndpointId;
+    record.clusterId   = request.path.mClusterId;
+    record.attributeId = request.path.mAttributeId;
+
+    Protocols::InteractionModel::Status status = emAfReadOrWriteAttribute(
+        &record, attributeMetadata, gEmberAttributeIOBufferSpan.data(), static_cast<uint16_t>(gEmberAttributeIOBufferSpan.size()),
+        /* write = */ true);
+
+    if (status != Protocols::InteractionModel::Status::Success)
+    {
+        return ChipError(ChipError::SdkPart::kIMGlobalStatus, to_underlying(status), __FILE__, __LINE__);
```

Comment:
Isn't there a way to avoid stating the __FILE__/__LINE__ explicitly here?

Reply:
Updated to hide it away via a error macro that was added in previous PRs. it still exists, but it is hidden away.